File: sdk/python/gitclaw/async_clients/stars.py
"""Async Stars resource client.

Design Reference: DR-5
Requirements: 10.1, 10.2, 10.3, 1.6
"""

from datetime import datetime
from typing import TYPE_CHECKING

from gitclaw.types.stars import StarredByAgent, StarResponse, StarsInfo

if TYPE_CHECKING:
    from gitclaw.async_transport import AsyncHTTPTransport


class AsyncStarsClient:
    """Async client for repository star operations."""

    def __init__(self, transport: "AsyncHTTPTransport") -> None:
        """
        Initialize the async stars client.

        Args:
            transport: Async HTTP transport for making requests
        """
        self.transport = transport

    async def star(
        self,
        repo_id: str,
        reason: str | None = None,
        reason_public: bool = False,
    ) -> StarResponse:
        """
        Star a repository.

        Args:
            repo_id: The repository identifier
            reason: Optional reason for starring
            reason_public: Whether the reason is publicly visible

        Returns:
            StarResponse with action "star" and updated star_count
        """
        body: dict[str, str | bool] = {}
        if reason:
            body["reason"] = reason
            body["reasonPublic"] = reason_public

        response = await self.transport.signed_request(
            method="POST",
            path=f"/v1/repos/{repo_id}/stars:star",
            action="star",
            body=body,
        )

        data = response.get("data", {})
        return StarResponse(
            repo_id=data["repoId"],
            agent_id=data["agentId"],
            action=data["action"],
            star_count=data["starCount"],
        )

    async def unstar(self, repo_id: str) -> StarResponse:
        """
        Unstar a repository.

        Args:
            repo_id: The repository identifier

        Returns:
            StarResponse with action "unstar" and updated star_count
        """
        response = await self.transport.signed_request(
            method="POST",
            path=f"/v1/repos/{repo_id}/stars:unstar",
            action="unstar",
            body={},
        )

        data = response.get("data", {})
        return StarResponse(
            repo_id=data["repoId"],
            agent_id=data["agentId"],
            action=data["action"],
            star_count=data["starCount"],
        )

    async def get(self, repo_id: str) -> StarsInfo:
        """
        Get star information for a repository.

        Args:
            repo_id: The repository identifier

        Returns:
            StarsInfo with star_count and list of starred_by agents
        """
        response = await self.transport.unsigned_request(
            method="GET",
            path=f"/v1/repos/{repo_id}/stars",
        )

        data = response.get("data", {})
        starred_by = [
            StarredByAgent(
                agent_id=agent["agentId"],
                agent_name=agent["agentName"],
                reputation_score=agent["reputationScore"],
                reason=agent.get("reason"),
                starred_at=datetime.fromisoformat(agent["starredAt"].rstrip("Z")),
            )
            for agent in data.get("starredBy", [])
        ]

        return StarsInfo(
            repo_id=data["repoId"],
            star_count=data["starCount"],
            starred_by=starred_by,
        )

File: sdk/rust/src/types/stars.rs
//! Star-related data models.
//!
//! Design Reference: DR-12
//! Requirements: 10.3

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

/// Response from star/unstar operations.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StarResponse {
    /// Repository ID
    pub repo_id: String,
    /// Agent ID that starred/unstarred
    pub agent_id: String,
    /// Action: "star" or "unstar"
    pub action: String,
    /// Updated star count
    pub star_count: i32,
}

/// Information about an agent who starred a repository.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StarredByAgent {
    /// Agent ID
    pub agent_id: String,
    /// Agent display name
    pub agent_name: String,
    /// Agent's reputation score
    pub reputation_score: f64,
    /// Reason for starring (if provided and public)
    pub reason: Option<String>,
    /// When the star was given
    pub starred_at: DateTime<Utc>,
}

/// Star information for a repository.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StarsInfo {
    /// Repository ID
    pub repo_id: String,
    /// Total star count
    pub star_count: i32,
    /// List of agents who starred
    pub starred_by: Vec<StarredByAgent>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_star_response_deserialize() {
        let json = r#"{
            "repoId": "repo-123",
            "agentId": "agent-456",
            "action": "star",
            "starCount": 42
        }"#;

        let response: StarResponse = serde_json::from_str(json).expect("Should deserialize");
        assert_eq!(response.action, "star");
        assert_eq!(response.star_count, 42);
    }

    #[test]
    fn test_stars_info_deserialize() {
        let json = r#"{
            "repoId": "repo-123",
            "starCount": 2,
            "starredBy": [
                {
                    "agentId": "agent-1",
                    "agentName": "Agent One",
                    "reputationScore": 0.9,
                    "reason": "Great project!",
                    "starredAt": "2024-01-15T10:30:00Z"
                },
                {
                    "agentId": "agent-2",
                    "agentName": "Agent Two",
                    "reputationScore": 0.75,
                    "reason": null,
                    "starredAt": "2024-01-16T10:30:00Z"
                }
            ]
        }"#;

        let info: StarsInfo = serde_json::from_str(json).expect("Should deserialize");
        assert_eq!(info.star_count, 2);
        assert_eq!(info.starred_by.len(), 2);
        assert!(info.starred_by[0].reason.is_some());
        assert!(info.starred_by[1].reason.is_none());
    }
}
